zips/zip-0032.rst

```diff
@@ -513,43 +524,119 @@ valid diversifiers.
+derivation paths of the application protocol's key tree in the corresponding ZIP is
+substantially complete, before public deployment of software or hardware using this
+mechanism. The ZIP process allows for subsequent updates and corrections.
 
 Let $\mathsf{ContextString}$ be a globally-unique non-empty sequence of at most 252 bytes
 that identifies the desired context.
 
 We instantiate the hardened key generation process with the following constants:
 
-- $\mathsf{Arbitrary.MKGDomain} = \texttt{“ZcashArbitraryKD”}$
-- $\mathsf{Arbitrary.CKDDomain} = \mathtt{0xAB}$
+- $\mathsf{Registered.MKGDomain} = \texttt{“ZIPRegistered\_KD”}$
+- $\mathsf{Registered.CKDDomain} = \mathtt{0xAC}$
+
+Registered subtree root key generation
+--------------------------------------
+
+Let $S$ be a seed byte sequence of a chosen length, which MUST be at least 32 and at most 252 bytes.
+
+The registered subtree root is obtained as:
 
-Arbitrary master key generation
+$\mathsf{RegKD}(\mathsf{ZipNumber}, \mathsf{ContextString}, \mathsf{S})$ $\rightarrow (\mathsf{sk}_{subtree}, \mathsf{c}_{subtree})$ :
+
+- Let $(\mathsf{sk}_{m}, \mathsf{c}_{m}) = \mathsf{MKGh}^\mathsf{Registered}([\mathsf{length}(\mathsf{ContextString})]\,||\,\mathsf{ContextString}\,||\,[\mathsf{length}(S)]\,||\,S)$.
+- Return $\mathsf{CKDh}^\mathsf{Registered}((\mathsf{sk}_{m}, \mathsf{c}_{m}), \mathsf{ZipNumber} + 2^{31}, [\,], \mathsf{false})$.
+
+Note: The intermediate key $(\mathsf{sk}_{m}, \mathsf{c}_{m})$ would grant the ability to derive the
+subtree root for application protocols defined in *any* ZIP using the same $\mathsf{ContextString}$
+and seed. This is a potentially dangerous scope of grant since it cannot be known what future
+protocols will use this mechanism; therefore, $(\mathsf{sk}_{m}, \mathsf{c}_{m})$ SHOULD NOT be
+used or stored directly without careful consideration of security consequences.
+
+Registered child key derivation
 -------------------------------
 
+As well as the integer child index $i$, the child key derivation function defined here supports a
+byte sequence $\mathsf{tag}$; each pair $(i, \mathsf{tag})$ produces a different child key. If an
+explicit tag is not required, $\mathsf{tag}$ is set to the empty byte sequence.
+
+Note: The tag MUST use an unambiguous encoding within the given context. It is RECOMMENDED to use
+a prefix-free encoding, which may require the use of length fields if multiple fields need to be
+encoded.
+
+$\mathsf{CKDreg}((\mathsf{sk}_{par}, \mathsf{c}_{par}), i, \mathsf{tag})$ $\rightarrow (\mathsf{sk}_{child}, \mathsf{c}_{child})$ :
+
+- Return $\mathsf{CKDh}^\mathsf{Arbitrary}((\mathsf{sk}_{par}, \mathsf{c}_{par}), i, \mathsf{tag}, \mathsf{false})$.
+
+Full-width child cryptovalue derivation
+---------------------------------------
+
+If the application protocol requires a 64-byte cryptovalue (for example, to avoid an entropy
+bottleneck in its subsequent operations), then a similar process to `Registered child key derivation`_
```

Comment:
I think that describing this as similar to `Registered child key derivation` doesn't illuminate the purpose here. More clarity or a more concrete example of an application protocol where this is applicable would be appreciated.

Reply:
We don't have an example of a protocol that uses this. I think it's sufficiently clear how it would be used.